opencog/atoms/distvalue/DistributionalValue.cc

```diff
@@ -0,0 +1,455 @@
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program; if not, write to:
+ * Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+#include <float.h>
+#include <math.h>
+#include <stdio.h>
+
+#include <opencog/util/numeric.h>
+
+#include <opencog/atoms/distvalue/DistributionalValue.h>
+#include <opencog/atomspace/AtomSpace.h>
+
+#include <boost/range/combine.hpp>
+#include <iomanip>
+
+using namespace opencog;
+
+count_t DistributionalValue::DEFAULT_K = 800.0;
+
+DistributionalValue::DistributionalValue()
+	: Value(DISTRIBUTIONAL_VALUE)
+{}
+
+DistributionalValue::DistributionalValue(const DVCounter &dvctr)
+	: Value(DISTRIBUTIONAL_VALUE)
+{
+	for (auto elem : dvctr)
+	{
+		if (elem.second <= 0)
+			throw std::invalid_argument("DV cant have an element with count 0.");
+	}
+	_value = dvctr;
+}
+
+//Create a DV from the Parameters of a SimpleTV
+//Not recommended as it results in a DV with only 1 singleton set
+//which in some calculations is unusalbe as the chance of overlaps in the
+//Sets/Intervals is small
+DistributionalValue::DistributionalValue(double mode,double conf)
+	: Value(DISTRIBUTIONAL_VALUE)
+{
+	confidence_t cf = std::min(conf, 0.9999998);
+	double count = (DEFAULT_K * cf / (1.0 - cf));
+	//DV Interval with count 0 is undefined
+	count = std::max(count, 0.0000002);
+	DVKey k{Interval{mode}};
+	_value[k] = count;
+}
+
+DistributionalValuePtr DistributionalValue::createDV(const DVCounter &dvctr)
+{
+	return std::make_shared<const DistributionalValue>(dvctr);
+}
+
+DistributionalValuePtr DistributionalValue::createDV(double mode,
+                                                     double conf)
+{
+	return std::make_shared<const DistributionalValue>(mode,conf);
+}
+
+DistributionalValuePtr
+DistributionalValue::UniformDistributionalValue(const DVKey &k,int c)
+{
+	DVCounter dvctr;
+	dvctr[k] = c;
+	return createDV(dvctr);
+}
+
+
+DistributionalValuePtr
+DistributionalValue::UniformDistributionalValue(const DVKeySeq &ks,int c)
+{
+	DVCounter dvctr;
+	for (auto k : ks)
+	{
+		dvctr[k] = c;
+	}
+	return createDV(dvctr);
+}
+
+DistributionalValuePtr DistributionalValue::TRUE_TV()
+{
+	DVKey v1{Interval{1.0}};
+	DVKey v2{Interval{0.0}};
+	DVCounter dvc;
+	dvc[v1] = 1;
+	dvc[v2] = 0;
+	return std::make_shared<const DistributionalValue>(dvc);
+}
+DistributionalValuePtr DistributionalValue::FALSE_TV()
+{
+	DVKey v1{Interval{1.0}};
+	DVKey v2{Interval{0.0}};
+	DVCounter dvc;
+	dvc[v1] = 0;
+	dvc[v2] = 1;
+	return std::make_shared<const DistributionalValue>(dvc);
+}
+DistributionalValuePtr DistributionalValue::DEFAULT_TV()
+{
+	DVKey v1{Interval{1.0}};
+	DVKey v2{Interval{0.0}};
+	DVCounter dvc;
+	dvc[v1] = 0;
+	dvc[v2] = 0;
+	return std::make_shared<const DistributionalValue>(dvc);
+}
+
+bool DistributionalValue::is_uniform() const
+{
+	double val = _value.begin()->second;
+	for (auto p : _value)
+	{
+		if (val != p.second)
+		{
```

Comment:
val is a `double`. This won't work. You have to `is_approx_equal_ulp` or something like that.

Furthermore, a distribution will NEVER be flat to within an ulp or two. A distribution is considered "flat" when the variation between bins is less than sqrt(N)/N where N is the average count per bin.   Also -- I'm missing a factor of sqrt(pi/2)  Also: when computing the "average count per bin" one must NOT include the counts from the overflow bins in the average!!

This is subtle stuff, and easy to get wrong, if you just wing it.  Should be coding this with a book or good reference by your side.